// smooth/lib/client/_error.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ErrorPage;

var _core = require("@emotion/core");

/** @jsx jsx */
function ErrorPage({
  error
}) {
  return (0, _core.jsx)("div", null, (0, _core.jsx)("div", {
    css: _core.css`
          color: #000;
          background: #fff;
          font-family: -apple-system, BlinkMacSystemFont, Roboto, 'Segoe UI',
            'Fira Sans', Avenir, 'Helvetica Neue', 'Lucida Grande', sans-serif;
          height: 90vh;
          text-align: center;
          display: flex;
          flex-direction: column;
          align-items: center;
          justify-content: center;
        `
  }, (0, _core.jsx)("div", null, (0, _core.jsx)("h1", {
    css: _core.css`
              display: inline-block;
              border-right: 1px solid rgba(0, 0, 0, 0.3);
              margin: 0;
              margin-right: 20px;
              padding: 10px 24px 10px 0;
              font-size: 24px;
              font-weight: 500;
              vertical-align: top;
            `
  }, error.statusCode || error.name), (0, _core.jsx)("div", {
    css: _core.css`
              display: inline-block;
              text-align: left;
              line-height: 50px;
              height: 50px;
              vertical-align: middle;
            `
  }, (0, _core.jsx)("h2", {
    css: _core.css`
                font-size: 14px;
                font-weight: normal;
                line-height: inherit;
                margin: 0;
                padding: 0;
              `
  }, error.message))), (0, _core.jsx)("code", {
    css: _core.css`
            padding-top: 20px;
            max-height: 400px;
            overflow: auto;
            max-width: 90%;
            text-align: left;
          `
  }, (0, _core.jsx)("pre", null, error.stack))));
}
