// smooth/lib/client/html.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = HTML;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/no-danger, jsx-a11y/html-has-lang */
function HTML(props) {
  return _react.default.createElement("html", props.htmlAttributes, _react.default.createElement("head", null, _react.default.createElement("meta", {
    charSet: "utf-8"
  }), _react.default.createElement("meta", {
    httpEquiv: "x-ua-compatible",
    content: "ie=edge"
  }), _react.default.createElement("meta", {
    name: "viewport",
    content: "width=device-width, initial-scale=1, shrink-to-fit=no"
  }), props.headComponents), _react.default.createElement("body", props.bodyAttributes, props.preBodyComponents, _react.default.createElement("div", {
    key: "body",
    id: "___smooth",
    dangerouslySetInnerHTML: {
      __html: props.body
    }
  }), props.postBodyComponents));
}
