// smooth/lib/client/Routes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Routes;

var _react = _interopRequireDefault(require("react"));

var _router = require("../router");

var _Page = _interopRequireWildcard(require("../page/Page"));

var _HiddenRouter = require("../router/HiddenRouter");

var _CoreRouter = require("../router/CoreRouter");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const pages = (0, _Page.getPages)();

function Routes({
  match: {
    url
  }
}) {
  const routes = _react.default.createElement(_router.Switch, null, pages.map((page, index) => _react.default.createElement(_router.Route, {
    key: index,
    path: `${url}${page.routePath}`.replace(/\/\//, '/'),
    render: ({
      history,
      match,
      location
    }) => _react.default.createElement(_Page.default, {
      indexUrl: `${url}${page.indexPath}`,
      page: page,
      history: history,
      match: match,
      location: location
    })
  })));

  return _react.default.createElement(_CoreRouter.CoreRouter, null, routes, _react.default.createElement(_HiddenRouter.HiddenRouter, null, routes));
}
