// smooth/lib/router/Router.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useRouter = useRouter;
Object.defineProperty(exports, "Router", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.Router;
  }
});
Object.defineProperty(exports, "BrowserRouter", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.BrowserRouter;
  }
});
Object.defineProperty(exports, "StaticRouter", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.StaticRouter;
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.Route;
  }
});
Object.defineProperty(exports, "Redirect", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.Redirect;
  }
});
Object.defineProperty(exports, "Prompt", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.Prompt;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.Switch;
  }
});
Object.defineProperty(exports, "generatePath", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.generatePath;
  }
});
Object.defineProperty(exports, "matchPath", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.matchPath;
  }
});
Object.defineProperty(exports, "withRouter", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.withRouter;
  }
});

var _react = require("react");

var _reactRouterDom = require("react-router-dom");

function useRouter() {
  return (0, _react.useContext)(_reactRouterDom.__RouterContext);
}
