// smooth/lib/config/plugins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parsePlugins = parsePlugins;

var _path = _interopRequireDefault(require("path"));

var _require = _interopRequireDefault(require("../babel/require"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function hasFile(plugin, name) {
  try {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    require.resolve(_path.default.join(plugin.resolve, name));

    return true;
  } catch (error) {
    return false;
  }
}

function loadFile(plugin, name) {
  try {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    return (0, _require.default)(_path.default.join(plugin.resolve, name));
  } catch (error) {
    return null;
  }
}

function parsePlugins(plugins) {
  return plugins.map(plugin => {
    if (typeof plugin === 'string') {
      plugin = {
        resolve: plugin
      };
    } // eslint-disable-next-line global-require, import/no-dynamic-require


    const node = hasFile(plugin, 'smooth-node');
    const browser = hasFile(plugin, 'smooth-browser');
    const nodePlugin = loadFile(plugin, 'smooth-node');
    const options = plugin.options || {};
    return {
      node,
      plugin: nodePlugin,
      browser,
      resolve: plugin.resolve,
      options: nodePlugin && nodePlugin.resolveOptions ? nodePlugin.resolveOptions(options) : options
    };
  });
}
