// smooth/lib/plugin/nodeHooks/onCreateApolloServerConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onCreateApolloServerConfig = void 0;

var _node = require("../node");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const onCreateApolloServerConfig = config => ({
  apolloServerConfig: defaultApolloServerConfig
}) => {
  let apolloServerConfig = defaultApolloServerConfig;
  (0, _node.applyHook)(config, 'onCreateApolloServerConfig', {
    getConfig() {
      return apolloServerConfig;
    },

    actions: {
      setApolloServerConfig(additionalConfig) {
        apolloServerConfig = _extends({}, apolloServerConfig, {}, additionalConfig);
      },

      replaceApolloServerConfig(newConfig) {
        apolloServerConfig = newConfig;
      }

    }
  });
  return apolloServerConfig;
};

exports.onCreateApolloServerConfig = onCreateApolloServerConfig;
