// smooth/lib/plugin/nodeHooks/onRenderBody.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onRenderBody = void 0;

var _node = require("../node");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const onRenderBody = config => ({
  headComponents = [],
  htmlAttributes = {},
  bodyAttributes = {},
  preBodyComponents = [],
  postBodyComponents = [],
  bodyProps = {},
  pathname,
  requestId
}) => {
  function setHeadComponents(components) {
    headComponents = [...headComponents, ...components];
  }

  function setHtmlAttributes(attributes) {
    htmlAttributes = _extends({}, htmlAttributes, {}, attributes);
  }

  function setBodyAttributes(attributes) {
    bodyAttributes = _extends({}, bodyAttributes, {}, attributes);
  }

  function setPreBodyComponents(components) {
    preBodyComponents = [...preBodyComponents, ...components];
  }

  function setPostBodyComponents(components) {
    postBodyComponents = [...postBodyComponents, ...components];
  }

  function setBodyProps(props) {
    bodyProps = _extends({}, bodyProps, {}, props);
  }

  (0, _node.applyHook)(config, 'onRenderBody', {
    pathname,
    requestId,
    setHeadComponents,
    setHtmlAttributes,
    setBodyAttributes,
    setPreBodyComponents,
    setPostBodyComponents,
    setBodyProps
  });
  return {
    headComponents,
    htmlAttributes,
    bodyAttributes,
    preBodyComponents,
    postBodyComponents,
    bodyProps
  };
};

exports.onRenderBody = onRenderBody;
