// smooth/lib/plugin/node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.applyHook = applyHook;
exports.applyAsyncHook = applyAsyncHook;
exports.callApi = callApi;
Object.defineProperty(exports, "runHook", {
  enumerable: true,
  get: function () {
    return _common.runHook;
  }
});
Object.defineProperty(exports, "runAsyncHook", {
  enumerable: true,
  get: function () {
    return _common.runAsyncHook;
  }
});
Object.defineProperty(exports, "hasHook", {
  enumerable: true,
  get: function () {
    return _common.hasHook;
  }
});

var _graphql = require("../graphql");

var _common = require("./common");

/* eslint-disable no-restricted-syntax */
function applyHook(config, ...args) {
  const defaultOptions = {
    config,
    types: _graphql.types
  };
  return (0, _common.createApplyHook)(config.plugins, defaultOptions)(...args);
}

async function applyAsyncHook(config, ...args) {
  const defaultOptions = {
    config,
    types: _graphql.types
  };
  return (0, _common.createApplyAsyncHook)(config.plugins, defaultOptions)(...args);
}

async function callApi(plugins, api, request) {
  for (const plugin of plugins) {
    if (plugin.plugin[api]) {
      return plugin.plugin[api]({
        request
      }, plugin.options);
    }
  }

  return null;
}
